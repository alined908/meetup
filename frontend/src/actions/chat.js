import * as types from "../constants/action-types"
import {axiosClient} from '../accounts/axiosClient'

export const getRoom = (uri) => async dispatch => {
    dispatch({type: types.GET_ROOMS_REQUEST})
    try {
        const response = await axiosClient.get(
            `/api/chats/${uri}/`, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
            }}
        )
        dispatch({type: types.GET_ROOMS_SUCCESS, payload: response.data})
    } catch(e){
        dispatch({type: types.GET_ROOMS_ERROR, payload: "Unable to get chat room."})
    }
}   

export const getRooms = () => async dispatch => {
    dispatch({type: types.GET_ROOMS_REQUEST})
    try {
        const response = await axiosClient.get(
            "/api/chats/", {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
            }}
        )
        dispatch({type: types.GET_ROOMS_SUCCESS, payload: response.data})
    } catch(e){
        dispatch({type: types.GET_ROOMS_ERROR, payload: "Unable to get chat rooms."})
    }
}

export const getMessages = (room) => async dispatch => {
    dispatch({type: types.GET_MESSAGES_REQUEST})
    try {
        const response = await axiosClient.get(
            `/api/chats/${room}/messages/`, {headers: {
            "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        dispatch({type: types.GET_MESSAGES_SUCCESS, payload: response.data})
    } catch(e){
        dispatch({type: types.GET_MESSAGES_ERROR, payload: "Unable to get messages."})
    }
}

export const getMoreMessages = (room, last_message_id) => async dispatch => {
    dispatch({type: types.GET_MORE_MESSAGES_REQUEST})
    try {
        const response = await axiosClient.get(
            `/api/chats/${room}/messages/`, {params: {last: last_message_id}, headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        dispatch({type: types.GET_MORE_MESSAGES_SUCCESS, payload: response.data})
    } catch(e){
        dispatch({type: types.GET_MORE_MESSAGES_ERROR, payload: "Unable to get more messages."})
    }

}

export const setActiveRoom = (room) => {
    return {
        type: types.SET_ACTIVE_ROOM,
        payload: {"uri": room}
    }
}

export const removeActiveRoom = () => {
    return {type: types.REMOVE_ACTIVE_ROOM}
}

export const updateRoom = (event) => async dispatch => {
    dispatch({type: types.UPDATE_ROOM, payload: event.message})
}

export const setTypingValue = (value) => {
    return {
        type: types.SET_TYPING_VALUE,
        payload: value
    }
}

export const addMessage = (message) => async dispatch => {
    dispatch({type: types.ADD_MESSAGE, payload: {message}})    
}