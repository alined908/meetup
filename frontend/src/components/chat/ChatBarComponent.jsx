import React, {Component} from 'react';
import {ContactComponent} from "../components";
import {connect} from 'react-redux';
import {IconButton, Tooltip} from '@material-ui/core'
import {Event as EventIcon, Person as PersonIcon, Search as SearchIcon, Error as ErrorIcon} from '@material-ui/icons';
const opposite = {"friend": "meetup", "meetup": "friend"}

class ChatBarComponent extends Component {
    constructor(props){
        super(props)
        this.state = {
            rooms: [],
            filteredRooms: [],
            searchedRooms: [],
            liveRooms: [],
            filters: {
                "friend": false, 
                "meetup": false
            },
            isFiltered: false,
            isSearched: false,
            searchInput: ''
        }
    }

    componentDidUpdate(prevProps) {
        if (this.props.rooms !== prevProps.rooms) {
            this.setState({rooms: this.props.rooms, liveRooms: this.props.rooms})
        }
    }

    filterRooms = (type) => {
        var rooms = this.state.isSearched ? this.state.searchedRooms : this.state.rooms;
        var isFiltered;
        var newRooms; 

        if (!this.state.filters[type]) {
            if (!this.state.filters[opposite[type]]){
                newRooms = rooms.filter(room => (type === "meetup" ? room.meetup === null : room.friendship === null))
            } else {
                newRooms = []
            }
            isFiltered = true
        } else {
            if (!this.state.filters[opposite[type]]){
                newRooms = rooms
                isFiltered = false
            } else {
                newRooms = rooms.filter(room => (type === "meetup" ?  room.friendship === null : room.meetup === null))
                isFiltered = true
            }
        }

        var filters = {...this.state.filters}
        filters[type] = !filters[type]
        this.setState({filters, isFiltered, filteredRooms: newRooms, liveRooms: newRooms})
    }

    handleSearchInputChange = (e) => {
        var filter = e.target.value;
        var rooms = this.state.isFiltered ? this.state.filteredRooms : this.state.rooms;
        var isSearched = e.target.value.length === 0 ? false : true;
        var newRooms;

        newRooms = rooms.filter((room) => {
            let meetupCriteria = false;
            if (room.meetup !== null) {
                meetupCriteria = room.name.toLowerCase().includes(filter.toLowerCase())
            }

            let friendCriteria = false;
            if (room.friendship !== null){
                const user = this.props.user
                let friend;
                for (var key in room.members){
                    if (key !== user.id.toString()){
                        friend = room.members[key]
                    }
                }
                const friendName = friend.first_name + " " + friend.last_name

                friendCriteria = (friendName.toLowerCase().includes(filter.toLowerCase())) || (friend.email.toLowerCase().includes(filter.toLowerCase()))
            }

            return meetupCriteria || friendCriteria
        })

        this.setState({searchInput: filter, isSearched, searchedRooms: newRooms, liveRooms: newRooms})
    }

    render (){
        const rooms = this.state.liveRooms

        return (
            <div className="chat-bar elevate">
                {!this.props.isRoomsInitialized && <div>Initializing Contacts...</div>}
                <div className="chat-bar-top">
                    <div>Contacts</div>
                    <div>
                        <Tooltip title="Friends">
                            <IconButton color={this.state.filters["friend"] ? "default" : "primary"} onClick={() => this.filterRooms("friend")} edge="end">
                                <PersonIcon/>
                            </IconButton>
                        </Tooltip>
                        <Tooltip title="Meetups">
                            <IconButton color={this.state.filters["meetup"] ? "default" : "primary"} onClick={() => this.filterRooms("meetup")}>
                                <EventIcon/>
                            </IconButton>
                        </Tooltip>
                    </div>
                </div>
                <div className="chat-bar-contacts">
                    {(this.props.isRoomsInitialized && rooms.length === 0) && 
                        <div className="chat-bar-none">
                            <ErrorIcon style={{color: "rgb(255, 212, 96)"}}/> &nbsp; No chat rooms match criteria!
                        </div>
                    }
                    {this.props.isRoomsInitialized && rooms.map((room) => 
                        <ContactComponent user={this.props.user} key={room.id} room={Object.values(room)}></ContactComponent>
                    )}
                </div>
                <div className="chat-bar-search">
                    <SearchIcon/>
                    <input className="chat-input" type="text" placeholder="Search Contacts..." value={this.state.searchInput} onChange={this.handleSearchInputChange}></input>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        user: state.user.user,
        isRoomsInitialized: state.chat.isRoomsInitialized
    }
}

export default connect(mapStateToProps)(ChatBarComponent)