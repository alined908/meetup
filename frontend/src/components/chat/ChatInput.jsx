import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import 'emoji-mart/css/emoji-mart.css'
import { Picker } from 'emoji-mart'
import { Button } from '@material-ui/core'
import styles from '../../styles/chat.module.css'

class ChatInput extends Component {
    constructor(props){
        super(props)
        this.state = {
            textValue: "",
            showEmojis: false
        }
        this.emojiPicker = React.createRef();
        this.showEmojis = this.showEmojis.bind(this);
    }
    
    // Add event listener for emoji picker outside click. 
    componentDidMount(){
        document.addEventListener('mousedown', this.handleOutsideClick, false);
    }

    // Remove event listener for emoji picker outside click. 
    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleOutsideClick, false);
    }

    // Send message if send button is clicked. 
    handleClick = (e) => {
        this.sendMessage();
    }

    // Send message if enter button is touched. 
    handleSubmit = (e) => {
        if (e.key === "Enter"){
            e.preventDefault();
            this.sendMessage();
        }
    }

    // Change text input value. 
    handleChange = (e) => {
        this.setState({textValue: e.target.value});    
    }

    // Send message to chat room websocket and reset chat input value. 
    sendMessage = () => {
        const messageObject = {
            from: this.props.user.id, 
            text: this.state.textValue, 
            room: this.props.room.uri
        };
        this.setState({textValue: ""}, 
            () => this.props.socket.newChatMessage(messageObject)
        );
    }

    // Add emoji to text input.
    addEmoji = e => {
        let emoji = e.native;
        this.setState({
            textValue: this.state.textValue + emoji
        });
    }

    // Toggle emoji picker visibility.
    showEmojis = () => {
        this.setState({showEmojis: !this.state.showEmojis})
    }

    // Handle emoji picker outside click.
    handleOutsideClick = (e) => {
        try {
            let node = ReactDOM.findDOMNode(this.emojiPicker.current)
            if (!node.contains(e.target)) {
                this.setState({showEmojis: false})
            }   
        } catch(error){
            return null
        }
    }

    render(){
        return (
            <>
                <form className={styles.inputForm}>
                    <input className={styles.input}
                        type="text"
                        onChange={this.handleChange} 
                        onKeyPress={this.handleSubmit}
                        value={this.state.textValue} 
                        placeholder="Type a message..">
                    </input>
                </form>
                {this.state.showEmojis ?
                    <div className={styles.emojiPickerWrapper}>
                        <div className={styles.emojiIcon} onClick={this.showEmojis}>
                            {String.fromCodePoint(0x1F62D)}
                        </div>
                        <div className={`${styles.emojiPicker} elevate`}>
                            <Picker 
                                ref={this.emojiPicker} 
                                title='Pick your emojiâ€¦' 
                                emoji='point_up' 
                                sheetSize={20} 
                                emojiSize={20} 
                                onSelect={this.addEmoji}
                            />
                        </div>
                    </div>:
                    <div className={styles.emojiIcon} onClick={this.showEmojis}>
                        {String.fromCodePoint(0x1f60a)}
                    </div>
                } 
                <div>
                    <Button 
                        disabled={this.state.textValue.length === 0}
                        onClick={() => this.handleClick()} 
                        className={styles.sendButton}
                    >
                        Send
                    </Button>
                </div>
            </>
        )
    }
}

export default ChatInput