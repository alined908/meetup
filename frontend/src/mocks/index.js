import moment from 'moment'

//Category
export const category = {
    id: 1, 
    image: null,
    label: "Dessert", 
    api_label: "desserts"
}

//User
export const user = {
    id: 1, 
    email: "daniel@gmail.com", 
    first_name: "Daniel", 
    last_name: "Lee",
    avatar: null
}
export const user2 = {
    id: 2, 
    email: "example2@gmail.com", 
    first_name: "Bob", 
    last_name: "Jim", 
    avatar: null
}
export const user3 = {
    id: 3, 
    email: "example3@gmail.com", 
    first_name: "Bob", 
    last_name: "Jim", 
    avatar: null
}
export const settings = {
    radius: 25, 
    location: null, 
    latitude: null,
    longitude: null
}

export const userWithSettings = {
    ...user,
    settings: {...settings}
}
export const friend = {
    id: 1, 
    user: user2, 
    created_at: moment().format(), 
    chat_room: "something"
}
export const friends = [friend]
export const preference = {
    id: 1, 
    user, 
    category, 
    name: "Dessert", 
    ranking: 1, 
    timestamp: moment().format()
}
export const preferences = [preference]
export const friendInvite = {
    id: 1, 
    timestamp: moment().format(), 
    status: 1, 
    uri: "random", 
    sender: user, 
    receiver: user2
}

//Chat
export const message = {
    id: 1, 
    message: "hello", 
    timestamp: moment().format(), 
    room_id: 1, 
    sender: user, 
    is_notif: false
}
export const message2 = {
    id: 2, 
    message: "hello", 
    timestamp: moment().format(), 
    room_id: 1, 
    sender: user, 
    is_notif: true
}
export const messages = [message, message2]
export const members = {1: user, 2: user2}
export const room = {
    uri: {
        id: 1, 
        uri: "abc", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: 1, 
        meetup:null, 
        notifs: 0
    }
}
export const rooms = {
    friend: {
        id: 1, 
        uri: "abc", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: 1, 
        meetup:null, 
        notifs: 0
    },
    meetup: {
        id: 2, 
        uri: "xyz", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: null, 
        meetup:1, 
        notifs: 3
    }
}

//General
export const error = {
    message: "Something went wrong."
}

//Meetup
export const meetupMember1 = {
    1: {
        id: 1,
        user,
        ban: false,
        admin: true
    }
}

export const meetupMember2 = {
    2: {
        id: 2,
        user: user2,
        ban: false,
        admin: false
    }
}

export const meetupMember3 = {
    3: {
        id: 3,
        user: user3,
        ban: false,
        admin: false
    }
}

export const meetupMembers = {...meetupMember1, ...meetupMember2}

export const meetup = {
    uri: {
        id: 1, 
        name: "Meetup", 
        uri: "uri", 
        creator: user, 
        location: "location", 
        date: moment().format("YYYY-MM-DD"),
        members: meetupMembers,
        notifs: 0,
        public: true,
        categories: [category],
        latitude: 34.228754, 
        longitude: -118.2351192,
        events: {},
        isMeetupEventsInitialized : true
    }
}

export const meetups = {...meetup}

export const event = {
    1: {
        id: 1,
        meetup: 1,
        creator: user,
        title: "Dessert",
        start: moment().format(),
        end: moment().format(),
        chosen: null,
        categories: [category],
        options: {},
        price: "1, 2",
        distance: 16000,
        entries: {},
        random: true
    }
}
export const events = {...event}

export const option = {
    1 : {
        id: 1,
        score: 0,
        option: null,
        restaurant: {},
        votes: {},
        banned: false
    }
}

export const options = {...option}

export const tokens = {
    access: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoiYWNjZXNzIiwiZXhwIjoxNTg4MjI5OTEzLCJqdGkiOiIwOTg2YTg0MDA1ZGM0OGQ1ODEzYmIyNzE1OWViNGUyZCIsInVzZXJfaWQiOjEsInVzZXIiOnsiaWQiOjEsImVtYWlsIjoiZGFuaWVsQGdtYWlsLmNvbSIsImZpcnN0X25hbWUiOiJEYW5pZWwiLCJsYXN0X25hbWUiOiJMZWUiLCJhdmF0YXIiOm51bGwsInNldHRpbmdzIjp7InJhZGl1cyI6MjUsImxvY2F0aW9uIjpudWxsLCJsYXRpdHVkZSI6bnVsbCwibG9uZ2l0dWRlIjpudWxsfX19.FbffRthOKX5F0TGvqA5mZXuXUWNkUojzKexDGCoP518",
    refresh : "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoicmVmcmVzaCIsImV4cCI6MTU4ODM4Mjk0NiwianRpIjoiN2ZhNGY2MDE4OWM3NDA5ZmFjOTMwYmE4ZTRlMWFhMTMiLCJ1c2VyX2lkIjoxLCJ1c2VyIjp7ImlkIjoxLCJlbWFpbCI6ImRhbmllbEBnbWFpbC5jb20iLCJmaXJzdF9uYW1lIjoiRGFuaWVsIiwibGFzdF9uYW1lIjoiTGVlIiwiYXZhdGFyIjpudWxsLCJzZXR0aW5ncyI6eyJyYWRpdXMiOjI1LCJsb2NhdGlvbiI6bnVsbCwibGF0aXR1ZGUiOm51bGwsImxvbmdpdHVkZSI6bnVsbH19fQ.qqaY4Ye7rmyOKLbDS2EKz4SGWH0fwZ0MvrGP6NJ3IRE"
}