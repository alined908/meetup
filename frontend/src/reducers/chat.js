import {GET_ROOMS_REQUEST, GET_ROOMS_SUCCESS, GET_ROOMS_ERROR, 
    GET_MESSAGES_REQUEST, GET_MESSAGES_SUCCESS, GET_MESSAGES_ERROR, 
    GET_MORE_MESSAGES_REQUEST, GET_MORE_MESSAGES_SUCCESS, GET_MORE_MESSAGES_ERROR,
    CLEAR_STORE, REMOVE_ACTIVE_ROOM, ADD_ROOM, SET_ACTIVE_ROOM, SET_TYPING_VALUE, ADD_MESSAGE} from "../constants/action-types"

const defaultState = {
    rooms: {},
    messages: [],
    activeRoom: null,
    setTypingValue: "",
    isActiveRoomSet: false,
    isRoomsInitialized: false,
    isRoomsFetching: false,
    isMessagesInitialized: false,
    isMessagesFetching: false,
    isMoreMessagesFetching: false
}

export default function chatReducer(state = defaultState, action){
    switch(action.type){
        case GET_ROOMS_REQUEST:
            return {...state, isRoomsFetching: true}
        case GET_ROOMS_SUCCESS:
            return {...state, rooms: action.payload, isRoomsInitialized: true, isRoomsFetching: false}
        case GET_ROOMS_ERROR:
            return {...state, isRoomsFetching: false, errorMessage: action.payload.message}
        case GET_MESSAGES_REQUEST:
            return {...state, isMessagesFetching: true}
        case GET_MESSAGES_SUCCESS:
            return {...state, messages: action.payload, isMessagesInitialized: true, isMessagesFetching: false}
        case GET_MESSAGES_ERROR:
            return {...state, isMessagesFetching: false, errorMessage: action.payload.message}
        case GET_MORE_MESSAGES_REQUEST:
            return {...state, isMoreMessagesFetching: true}
        case GET_MORE_MESSAGES_SUCCESS:
            return {...state, messages: [...action.payload, ...state.messages], isMoreMessagesFetching: false}
        case GET_MORE_MESSAGES_ERROR:
            return {...state, errorMessage: action.payload.message, isMoreMessagesFetching: false}
        case ADD_ROOM:
            return {...state, rooms: state.rooms.concat([action.payload])}
        case SET_ACTIVE_ROOM:
            return {...state, activeRoom: action.payload.uri, isActiveRoomSet: true, isMessagesInitialized: false}
        case REMOVE_ACTIVE_ROOM:
            return {...state, activeRoom: null, isActiveRoomSet:false, isMessagesInitialized:false}
        case SET_TYPING_VALUE:
            return {...state, setTypingValue: action.payload}
        case ADD_MESSAGE:
            return {...state, setTypingValue: '', messages: [...state.messages, action.payload.message]}
        case CLEAR_STORE:
            return defaultState
        default:
            return state
    }
}