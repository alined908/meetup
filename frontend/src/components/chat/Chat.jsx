import React, { Component } from 'react'
import { ChatBar, ChatWindow } from "../components"
import { connect } from "react-redux";
import {getMessages, setActiveRoom, addMessage, removeActiveRoom, getRooms} from "../../actions"
import WebSocketService from "../../accounts/WebSocket"
import AuthenticationService from '../../accounts/AuthenticationService';
import { Helmet } from "react-helmet"
import styles from '../../styles/chat.module.css'

class Chat extends Component {
    constructor(props){
        super(props)
        this.state = {
            socket: new WebSocketService(),
            nonMobile: window.matchMedia("(min-width: 768px)").matches
        }
        this.state.socket.addChatCallbacks(this.props.getMessages, this.props.addMessage)
    }

    componentDidMount(){
        const handler = e => this.setState({nonMobile: e.matches});
        window.matchMedia("(min-width: 768px)").addListener(handler);
        this.props.getRooms()
        if("uri" in this.props.match.params){
            this.getRelevantInfo(this.props.match.params.uri)
        }
    }

    componentDidUpdate(prevProps){
        if (this.props.match.params.uri !== prevProps.match.params.uri){
            if (this.state.socket.exists()){
                this.state.socket.disconnect()
            }
            this.getRelevantInfo(this.props.match.params.uri)
        }
    }

    componentWillUnmount(){
        this.props.removeActiveRoom()
        this.state.socket.disconnect()
    }

    getRelevantInfo(uri) {
        const socket = this.state.socket
        const token = AuthenticationService.retrieveToken()
        const path = `/ws/chat/${this.props.match.params.uri}/`;
        socket.connect(path, token);
        this.setRoomInfo(uri)
    }

    setRoomInfo(uri){
        this.props.setActiveRoom(uri);
        this.props.getMessages(uri);
    }

    render(){
        const renderChatWindow = () => {
            if (this.props.isMessagesInitialized || this.props.isMessagesFetching) {
                return <ChatWindow
                            socket={this.state.socket} isMessagesInitialized={this.props.isMessagesInitialized} 
                            activeRoom={this.props.activeRoom} messages={this.props.messages}
                        />
            } else {
                return <ChatWindow socket={this.state.socket} activeRoom={null} messages={[]}/>
            }
        }

        return (
            <div className={`${styles.chat}` + (this.state.nonMobile ? "" : "chat-mobile")}>
                <Helmet>
                    <meta charSet="utf-8" />
                    <title>Chat</title>
                    <meta name="description" content="Phoodie Chat" />
                </Helmet>
                <ChatBar rooms={this.props.rooms}/>
                {renderChatWindow()}
            </div>
        )
    }
}

const mapDispatchToProps = {
    setActiveRoom,
    getMessages,
    getRooms,
    addMessage,
    removeActiveRoom
}

function mapStateToProps(state){
    return {
        isMessagesInitialized: state.chat.isMessagesInitialized,
        activeRoom: state.chat.activeRoom,
        messages: state.chat.messages,
        isActiveRoomSet: state.chat.isActiveRoomSet,
        rooms: Object.values(state.chat.rooms),
        isMessagesFetching: state.chat.isMessagesFetching
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Chat)
export {Chat as UnderlyingChat}