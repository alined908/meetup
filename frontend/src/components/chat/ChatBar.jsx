import React, {Component} from 'react';
import {Contact} from "../components";
import {connect} from 'react-redux';
import {IconButton, Tooltip, CircularProgress} from '@material-ui/core'
import {Event as EventIcon, Person as PersonIcon, Search as SearchIcon, Error as ErrorIcon} from '@material-ui/icons';
import PropTypes from 'prop-types';
import {chatRoomPropType, userPropType,} from "../../constants/prop-types"
import WebSocketService from "../../accounts/WebSocket"
import AuthenticationService from '../../accounts/AuthenticationService';
import {updateRoom} from '../../actions'
import styles from '../../styles/chat.module.css'

class ChatBar extends Component {
    constructor(props){
        super(props)
        this.state = {
            rooms: [],
            filteredRooms: [],
            searchedRooms: [],
            liveRooms: [],
            filters: {
                "friend": false, 
                "meetup": false
            },
            isFiltered: false,
            isSearched: false,
            searchInput: '',
            socket: new WebSocketService()
        }
        this.state.socket.addContactsCallbacks(this.props.updateRoom)
    }

    // Initialize socket for chat rooms.
    componentDidMount(){
        const socket = this.state.socket
        const token = AuthenticationService.retrieveToken()
        const path = `/ws/chat/rooms/`;
        socket.connect(path, token);
    }

    componentDidUpdate(prevProps) {
        if (this.props.rooms !== prevProps.rooms) {
            this.setState({rooms: this.props.rooms, liveRooms: this.props.rooms})
        }
    }

    filterAndSearch = () => {
        const rooms = this.state.rooms
        const filter = this.state.searchInput
        const filterMeetup = this.state.filters["meetup"]
        const filterFriend = this.state.filters["friend"]
        let newRooms;

        newRooms = rooms.filter((room) => {
        
            let meetupCriteria = false;
            if (room.meetup !== null) {
                meetupCriteria = room.name.toLowerCase().includes(filter.toLowerCase()) && !filterMeetup
            }

            let friendCriteria = false;
            if (room.friendship !== null){
                const user = this.props.user
                let friend;
                for (let key in room.members){
                    if (key !== user.id.toString()){
                        friend = room.members[key]
                    }
                }
                const friendName = `${friend.first_name} ${friend.last_name}`

                friendCriteria = (
                    (friendName.toLowerCase().includes(filter.toLowerCase())) || 
                    (friend.email.toLowerCase().includes(filter.toLowerCase()))
                ) && !filterFriend
            }

            return meetupCriteria || friendCriteria
        })

        this.setState({liveRooms: newRooms})
    }

    filterRooms = (type) => {
        let filters = {...this.state.filters}
        filters[type] = !filters[type]
        this.setState({filters}, () => this.filterAndSearch())
    }

    handleSearchInputChange = (e) => {
        var filter = e.target.value;
        this.setState({searchInput: filter}, () => this.filterAndSearch())
    }

    render (){
        const rooms = this.state.liveRooms

        return (
            <div className={styles.bar}>
                <div className={styles.barTop}>
                    <div>Contacts</div>
                    <div>
                        <Tooltip title="Friends">
                            <IconButton 
                                color={this.state.filters["friend"] ? "default" : "primary"} 
                                onClick={() => this.filterRooms("friend")} edge="end"
                            >
                                <PersonIcon/>
                            </IconButton>
                        </Tooltip>
                        <Tooltip title="Meetups">
                            <IconButton 
                                color={this.state.filters["meetup"] ? "default" : "primary"} 
                                onClick={() => this.filterRooms("meetup")}
                            >
                                <EventIcon/>
                            </IconButton>
                        </Tooltip>
                    </div>
                </div>
                <div className={styles.contacts}>
                    {this.props.isRoomsFetching && 
                        <div className="loading">
                            <CircularProgress/>
                        </div>
                    }
                    {(this.props.isRoomsInitialized && rooms.length === 0) && 
                        <div className="no-entity">
                            <ErrorIcon style={{color: "rgb(255, 212, 96)"}}/>
                            <span className="no-entity-text">
                                No chat rooms match criteria!
                            </span>
                        </div>
                    }
                    {this.props.isRoomsInitialized && rooms.map((room) => 
                        <Contact 
                            key={room.id}
                            room={room}
                            user={this.props.user} 
                            currentRoom={this.props.currentRoom}
                        />
                    )}
                </div>
                <div className={styles.search}>
                    <SearchIcon/>
                    <input 
                        type="text"
                        className={styles.input} 
                        placeholder="Search Contacts..." 
                        value={this.state.searchInput} 
                        onChange={this.handleSearchInputChange}
                    />
                </div>
            </div>
        )
    }
}

ChatBar.propTypes = {
    rooms: PropTypes.arrayOf(
        chatRoomPropType
    ),
    user: userPropType,
    isRoomsInitialized: PropTypes.bool
}

const mapStateToProps = (state) => {
    return {
        user: state.user.user,
        isRoomsInitialized: state.chat.isRoomsInitialized,
        isRoomsFetching: state.chat.isRoomsFetching,
        currentRoom: state.chat.activeRoom
    }
}

const mapDispatchToProps = {
    updateRoom
}

export default connect(mapStateToProps, mapDispatchToProps)(ChatBar)
export {ChatBar as UnderlyingChatBar}