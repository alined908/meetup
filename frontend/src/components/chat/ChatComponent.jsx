import React, {Component} from 'react'
import ChatBarComponent from "./ChatBarComponent"
import ChatWindowComponent from "./ChatWindowComponent"
import {connect} from "react-redux";
import {getMessages, setActiveRoom, addMessage, removeActiveRoom} from "../../actions/chat"
import {getRooms} from '../../actions/chat';
import WebSocketService from "../../accounts/WebSocket"
import AuthenticationService from '../../accounts/AuthenticationService';

class ChatComponent extends Component {
    constructor(props){
        super(props)
        this.state = {
            socket: new WebSocketService(),
            nonMobile: window.matchMedia("(min-width: 768px)").matches
        }
        this.state.socket.addChatCallbacks(this.props.getMessages, this.props.addMessage)
    }

    componentDidMount(){
        const handler = e => this.setState({nonMobile: e.matches});
        window.matchMedia("(min-width: 768px)").addListener(handler);
        this.props.getRooms()
        if("uri" in this.props.match.params){
            this.getRelevantInfo(this.props.match.params.uri)
        }
    }

    componentDidUpdate(prevProps){
        if (this.props.match.params.uri !== prevProps.match.params.uri){
            if (this.state.socket.exists()){
                this.state.socket.disconnect()
            }
            this.getRelevantInfo(this.props.match.params.uri)
        }
    }

    componentWillUnmount(){
        this.props.removeActiveRoom()
        this.state.socket.disconnect()
    }

    getRelevantInfo(uri) {
        const socket = this.state.socket
        this.props.setActiveRoom(uri);
        this.props.getMessages(uri);
        var ws_scheme = window.location.protocol === "https:" ? "wss": "ws"
        const token = AuthenticationService.retrieveToken()
        const path = `${ws_scheme}://localhost:8000/ws/chat/${this.props.match.params.uri}/?token=${token}`;
        socket.connect(path);
    }

    render(){
        const renderChatWindow = () => {
            if (this.props.isMessagesInitialized) {
                return <ChatWindowComponent socket={this.state.socket} isMessagesInitialized={this.props.isMessagesInitialized} activeRoom={this.props.activeRoom} messages={this.props.messages}></ChatWindowComponent>
            } else {
                return <ChatWindowComponent socket={this.state.socket} activeRoom={null}></ChatWindowComponent>
            }
        }

        return (
            <div className={"chat " + (this.state.nonMobile ? "" : "chat-mobile")}>
                <ChatBarComponent rooms={this.props.rooms}/>
                {renderChatWindow()}
            </div>
        )
    }
}

const mapDispatchToProps = {
    setActiveRoom,
    getMessages,
    getRooms,
    addMessage,
    removeActiveRoom
}

function mapStateToProps(state){
    return {
        isMessagesInitialized: state.chat.isMessagesInitialized,
        activeRoom: state.chat.activeRoom,
        messages: state.chat.messages,
        isActiveRoomSet: state.chat.isActiveRoomSet,
        rooms: Object.values(state.chat.rooms)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ChatComponent)