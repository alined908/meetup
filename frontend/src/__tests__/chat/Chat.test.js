import React from 'react'
import {shallow, mount} from 'enzyme';
import {UnderlyingChat} from "../../components/chat/Chat"
import WebSocketService from "../../accounts/WebSocket";
import * as mocks from "../../mocks"

jest.mock("../../accounts/WebSocket")

const props = {
    messages: [...mocks.messages], 
    rooms: [], 
    match: {params: {}},
    setActiveRoom: jest.fn(),
    getMessages: jest.fn(),
    getRooms: jest.fn(),
    addMessage: jest.fn(),
    removeActiveRoom: jest.fn()
}

describe("Chat unit ", () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(query => ({
          matches: false,
          media: query,
          onchange: null,
          addListener: jest.fn(), // deprecated
          removeListener: jest.fn(), // deprecated
          dispatchEvent: jest.fn(),
        })),
    });

    const addChatCallbacks = jest.fn(), connect = jest.fn(), disconnect = jest.fn(),  exists = jest.fn()
    exists.mockReturnValue(true)
    WebSocketService.mockImplementation(() => {
        return {
            addChatCallbacks,
            connect,
            disconnect,
            exists
        }
    })

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ('renders without crashing given props', () => {
        const wrapper = shallow(<UnderlyingChat {...props}/>)
        expect(wrapper).toMatchSnapshot()
    })

    it ('renders and gets info for specific room', () => {
        const withUri = {...props, match: {params: {uri: "uri"}}}
        const wrapper = shallow(<UnderlyingChat {...withUri}/>)
        expect(props.getRooms.mock.calls.length).toBe(1)
        expect(props.setActiveRoom.mock.calls.length).toBe(1)
        expect(props.getMessages.mock.calls.length).toBe(1)
        expect(addChatCallbacks.mock.calls.length).toBe(1)
        expect(connect.mock.calls.length).toBe(1)
    })

    it ('handle componentDidUpdate', () => {
        const withUri = {...props, match: {params: {uri: "uri"}}}
        const wrapper = shallow(<UnderlyingChat {...withUri}/>)
        wrapper.setProps({match: {params: {uri: "something"}}})
        expect(exists.mock.calls.length).toBe(1)
        expect(disconnect.mock.calls.length).toBe(1)
        expect(props.setActiveRoom.mock.calls.length).toBe(2)
        expect(props.getMessages.mock.calls.length).toBe(2)
    })

    it ('handle componentwillunmount', () => {
        const wrapper = shallow(<UnderlyingChat {...props}/>)
        wrapper.unmount()
        expect(props.removeActiveRoom.mock.calls.length).toBe(1)
        expect(disconnect.mock.calls.length).toBe(1)
    })
})