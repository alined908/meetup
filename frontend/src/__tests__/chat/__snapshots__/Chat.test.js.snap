// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat unit  renders without crashing given props 1`] = `
<div
  className="chatchat-mobile"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <meta
      charSet="utf-8"
    />
    <title>
      Chat
    </title>
    <meta
      content="Phoodie Chat"
      name="description"
    />
  </HelmetWrapper>
  <Connect(ChatBar)
    rooms={Array []}
  />
  <Connect(ChatWindow)
    activeRoom={null}
    messages={Array []}
    socket={
      Object {
        "addChatCallbacks": [MockFunction] {
          "calls": Array [
            Array [
              [MockFunction],
              [MockFunction],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "connect": [MockFunction],
        "disconnect": [MockFunction],
        "exists": [MockFunction],
      }
    }
  />
</div>
`;
